### ISSUE TITLE
promtool: Add support for native histograms

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#12146]
### Proposal

Various parts of `promtool` require proper support for native histograms, for example the rules unit tests (but there are many more, once we have an overview, let's create a checklist here).

**Comment by ashutosh887** [comment]
@beorn7 I would like to work on this issue

**Comment by beorn7** [comment]
The rules unit tests now support native histograms. :tada:

**Comment by fatsheep9146** [comment]
After I have researched promtool, I think the all the subcommands already support native histogram or do not need to support native histogram are 

- check 
   - service-discovery: do not need support 
   - config: do not need support 
   - web-config: do not need support 
   - healthy: do not need support 
   - ready: do not need support 
   - rules: already native histogram
- query
   - instance: already support native histogram
   - range: already support native histogram
   - series: already support native histogram
   - labels: already support native histogram
- debug
   - pprof: do not need support 
- test
   - rule: already support native histogram
- promql: all subcommands support native histogram
- tsdb
	- analyze: already support native histogram
	- list: already support native histogram
	- create-blocks-from: do not need support 

So all the TODOs need to be done

- [ ] **check metrics**: This command only supports check metrics in text format for now, but the native histogram do not support text format.
- [ ] **debug metrics**: This commond only supports fetch metric int text format for now, but the native histogram do not support text format.
- [ ] **push metrics**: This command only supports load metrics in text format for now, but the native histogram do not support text format.
- [ ] **tsdb bench write**: This command only write float value for bench test, not write native histogram
- [ ] **tsdb dump**: This command only dump float type value, not native histogram value

@beorn7 

And I would love to help with it.

**Comment by beorn7** [comment]
Thanks for the great analysis. Note that there is also #12331 (which is more about creating a whole new sub-command and not really about adding support for native histograms to an existing command).

I think `tsdb analyze` still needs an update because we will now have chunks with more than 120 samples or deliberately less. That's not only caused by native histograms, but by other recent changes, too. See #12054 and #12055. So we need to handle that variable number of samples better, plus we should probably add some specific stuff for native histograms. A single histogram sample can have many buckets (or very few), and it would be good to see some stats about that, too. Seeing the size of chunks in bytes would be cool, too. So maybe have separate histograms for float chunks and histogram chunks, and each of them should have a histogram for sample count, chunk size in bytes, and for histogram number of buckets in the chunk (maybe even "represented buckets" vs. "populated buckets").

About the TODOs:

I agree that a lot depends on availability of text format support (which we'll probably only add to OpenMetrics rather than to the old text format – see #11265 ). So that means for `check metrics`, `debug metrics`, and `push metrics`, we simply have to wait and see what the OpenMetrics text format will look like. (I guess we could also utilize the protobuf format somehow for these commands, but I don't think the need is urgent enough. Better to get the OM text format rolling first.)

I think `tsdb bench write` would just be "nice to have". I'm not sure how much it is used right now. So I would only invest work into it if somebody explicitly requests it.

Which leaves us with `tsdb dump`: That would indeed be good to support native histograms. The dump uses pretty much an ad hoc format. That's a problem of its own, see discussion in #8281 in #8320 , but for now, it simplifies things here. You can essentially make something up (which could be using the `String` methods that are also used in the internal web UI).

So in summary, that's what you could work on right naw:
- `tsdb dump` should be pretty straight forward. 
-  `tsdb analyze` needs an update for variable chunk sizes anyway, plus more data for histograms. So that would be a bit more involved, but probably very helpful.

**Comment by fatsheep9146** [comment]
> Thanks for the great analysis. Note that there is also #12331 (which is more about creating a whole new sub-command and not really about adding support for native histograms to an existing command).
> 
> I think `tsdb analyze` still needs an update because we will now have chunks with more than 120 samples or deliberately less. That's not only caused by native histograms, but by other recent changes, too. See #12054 and #12055. So we need to handle that variable number of samples better, plus we should probably add some specific stuff for native histograms. A single histogram sample can have many buckets (or very few), and it would be good to see some stats about that, too. Seeing the size of chunks in bytes would be cool, too. So maybe have separate histograms for float chunks and histogram chunks, and each of them should have a histogram for sample count, chunk size in bytes, and for histogram number of buckets in the chunk (maybe even "represented buckets" vs. "populated buckets").
> 
> About the TODOs:
> 
> I agree that a lot depends on availability of text format support (which we'll probably only add to OpenMetrics rather than to the old text format – see #11265 ). So that means for `check metrics`, `debug metrics`, and `push metrics`, we simply have to wait and see what the OpenMetrics text format will look like. (I guess we could also utilize the protobuf format somehow for these commands, but I don't think the need is urgent enough. Better to get the OM text format rolling first.)
> 
> I think `tsdb bench write` would just be "nice to have". I'm not sure how much it is used right now. So I would only invest work into it if somebody explicitly requests it.
> 
> Which leaves us with `tsdb dump`: That would indeed be good to support native histograms. The dump uses pretty much an ad hoc format. That's a problem of its own, see discussion in #8281 in #8320 , but for now, it simplifies things here. You can essentially make something up (which could be using the `String` methods that are also used in the internal web UI).
> 
> So in summary, that's what you could work on right naw:
> 
> * `tsdb dump` should be pretty straight forward.
> * `tsdb analyze` needs an update for variable chunk sizes anyway, plus more data for histograms. So that would be a bit more involved, but probably very helpful.

Thanks! I will start working as you suggests

**Comment by fatsheep9146** [comment]
@beorn7 @dgl
I think this issue is only partly resolved, should be opened again.

**Comment by dgl** [comment]
@fatsheep9146 ack (I missed that GitHub had parsed your "partly fix #id" comment as "fixes" and so automatically closed this).

**Comment by fatsheep9146** [comment]
> @fatsheep9146 ack (I missed that GitHub had parsed your "partly fix #id" comment as "fixes" and so automatically closed this).

Thanks!

**Comment by fatsheep9146** [comment]
> So we need to handle that variable number of samples better.

I think for now we are not able to deduce the samples-per-chunk from the chunk file, could we just optimize the display format of histogram for compact analysis? 
Instead of displaying the percentage, we just display the absolute value of the sample count (but with a step size of 10 units.)

For example, for now the display format is like
```
Compaction analysis:
Fullness: Amount of samples in chunks (100% is 120 samples)
     10%:
     20%:
     30%:
     40%:
     50%:
     60%:
     70%:
     80%:
     90%: ####################
    100%: ################################################################################
```

We change to
```
     10:
     20:
     30:
     40:
     50:
     60:
     70:
     80:
     90: 
    100: ####################
    110:
    120: ################################################################################
```

@beorn7

**Comment by shaheislam** [comment]
Hi @beorn7 am I correct in understanding based off of this issue promtool is not capable of correctly running unit tests against expressions which leverage histograms?

**Comment by beorn7** [comment]
https://github.com/prometheus/prometheus/pull/12668 should have added that capability. So my expectation is that `promtool` will run unit tests involving native histograms just fine. If not, that would be a bug.

**Comment by beorn7** [comment]
To update the summary here: The remaining features are blocked on having a text format representation of native histograms. On the other hand, we probably don't need every single aspect supported in promtool before declaring native histograms a stable feature, so I'm demoting this to P3.
